Firestore/core/src/firebase/firestore/remote/grpc_stream.h

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_GRPC_STREAM_H
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_GRPC_STREAM_H
+
+#include <memory>
+#include <utility>
+
+#include <grpcpp/client_context.h>
+#include <grpcpp/generic/generic_stub.h>
+#include <grpcpp/support/byte_buffer.h>
+
+#include "Firestore/core/src/firebase/firestore/remote/buffered_writer.h"
+#include "Firestore/core/src/firebase/firestore/remote/grpc_operation.h"
+#include "Firestore/core/src/firebase/firestore/remote/grpc_queue.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+#include "absl/types/optional.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+namespace internal {
+class GrpcStreamDelegate;
+}
+
+/**
+ * A GRPC bidirectional stream that notifies the given `observer` about stream
+ * events.
+ *
+ * The stream has to be explicitly opened (via `Start`) before it can be used.
+ * The stream is always listening for new messages from the server. The stream
+ * can be used to send messages to the server (via `Write`); messages are queued
+ * and sent out one by one. Both sent and received messages are raw bytes;
+ * serialization and deserialization are left to the caller.
+ *
+ * The stream stores the generation number of the observer at the time of its
+ * creation; once observer increases its generation number, the stream will stop
+ * notifying it of events.
+ *
+ * The stream is disposable; once it finishes, it cannot be restarted.
+ *
+ * This class is a wrapper over `grpc::GenericClientAsyncReaderWriter`.
+ */
+class GrpcStream : public std::enable_shared_from_this<GrpcStream> {
```

Comment:
GRPC is a little inconsistent with naming, and consequently I'm not entirely confident how to call this. This class is largely a wrapper around `grpc::GenericClientAsyncReaderWriter`. On one hand, that "reader-writer" is defined in a file called `async_stream.h` and implements `ClientAsyncStreamingInterface`. On the other hand, it is created by a function called `PrepareCall`, so `GrpcCall` would be a plausible name (and perhaps better reflect its disposable nature).